File: visual_feedback/snapshotter/src/snapshot_filter.py
#!/usr/bin/env python

##    @package snapshotter

import roslib
import sys
roslib.load_manifest("snapshotter")
import rospy
import math
import tf
from tf.msg import tfMessage
import cv
from std_msgs.msg import String
from sensor_msgs.msg import Image,CameraInfo
from cv_bridge import CvBridge, CvBridgeError
from snapshotter.srv import *
from snapshotter.msg import *

class SnapshotFilter:

    ##   The constructor
    #    @param self The object pointer
    def __init__(self):
        input_topic = rospy.get_param("~input","defaultFilterInput")
        output_topic = rospy.get_param("~output","defaultFilterOutput")
        self.get_extended_params()
        self.name = rospy.get_name()
        self.bridge = CvBridge()
        self.input_topic = input_topic
        self.output_topic = output_topic
        self.output_pub = rospy.Publisher(self.output_topic,Snapshot)
        self.input_sub = rospy.Subscriber(self.input_topic,Snapshot,self.handle_input)
    
    def get_extended_params(self):
        #Do nothing
        return
    
    #Takes a snapshot as input, returns a filtered snapshot as output
    def handle_input(self,snapshot):
        try:
            input_cv_image = self.bridge.imgmsg_to_cv(snapshot.image, "bgr8")
            copy_input_cv_image = cv.CreateImage((input_cv_image.width,input_cv_image.height),8,3)
            cv.Copy(input_cv_image,copy_input_cv_image)
        except CvBridgeError, e:
            print e
        output_cv_image = self.image_filter(cv_image=input_cv_image,info=snapshot.info,copy=copy_input_cv_image)
        output_info = snapshot.info
        try:
            output_image = self.bridge.cv_to_imgmsg(output_cv_image, "bgr8")
        except CvBridgeError, e:
            print e
        filters = snapshot.filters
        filters.append(self.name.replace('/',''))
        output_snapshot = Snapshot(image=output_image,info=output_info,filters=filters)
        self.output_pub.publish(output_snapshot)
        print "%s output"%self.name
        
        
    #Abstract method, returns the altered image and info
    def image_filter(self,cv_image,info,copy=None):
        abstract
    
## Instantiate a new snapshotter node
def main(args):
    rospy.init_node("shapshot_filter")
    filt = SnapshotFilter()
    rospy.spin()

if __name__ == '__main__':
    args = sys.argv[1:]
    try:
        main(args)
    except rospy.ROSInterruptException: pass

File: visual_feedback/snapshotter/include/snapshot_filter.cpp
#ifndef SNAPSHOT_FILTER_CPP
#define SNAPSHOT_FILTER_CPP
#include "snapshot_filter.h"

SnapshotFilter::SnapshotFilter(){
    //Do nothing
}

SnapshotFilter::SnapshotFilter(ros::NodeHandle &n,ros::NodeHandle &glob) 
{
    ROS_INFO("constructor called");
    n_ = n;
    //n.getParam("name",name);
    name = ros::this_node::getName();
    n.getParam("input",input_topic);
    n.getParam("output",output_topic);
    get_extended_params();
    input_sub = glob.subscribe<snapshotter::Snapshot>(input_topic,1,&SnapshotFilter::handle_my_input,this);
    output_pub = glob.advertise<snapshotter::Snapshot>(output_topic,1);
    ROS_INFO("constructor finished");
}

SnapshotFilter::~SnapshotFilter()
{
    //Do nothing
}

void SnapshotFilter::get_extended_params(){};

IplImage* SnapshotFilter::image_filter(IplImage* img_ptr,sensor_msgs::CameraInfo* cam_info_ptr){
    return (IplImage*) img_ptr;
};

void SnapshotFilter::handle_my_input(const snapshotter::SnapshotConstPtr& snapshot_ptr)
{
    ROS_INFO("received");
    sensor_msgs::Image image = snapshot_ptr->image;
    sensor_msgs::ImagePtr img_ptr(new sensor_msgs::Image(image));
    sensor_msgs::CameraInfo cam_info = snapshot_ptr->info;
    
    IplImage *cv_image = NULL;
    
    try
    {
        cv_image = bridge_.imgMsgToCv(img_ptr, "bgr8");
    }
    catch (sensor_msgs::CvBridgeException error)
    {
        ROS_ERROR("error");
        return;
    }

    IplImage *output_cv_image = image_filter(cv_image,&cam_info);
    sensor_msgs::ImagePtr output_img_ptr;
    sensor_msgs::Image output_img;
    try
    {
        output_img_ptr = bridge_.cvToImgMsg(output_cv_image, "bgr8");
        output_img = *output_img_ptr;
    }
    catch (sensor_msgs::CvBridgeException error)
    {
        ROS_ERROR("error");
        return;
    }

    snapshotter::Snapshot snapshot;
    snapshot.image = output_img;
    snapshot.info = cam_info;

    ROS_INFO("Preparing to publish");
    output_pub.publish(snapshot);
    ROS_INFO("Published");

}

class CircleFilter : public SnapshotFilter{
    public:
    CircleFilter(ros::NodeHandle &n,ros::NodeHandle &glob) : SnapshotFilter(n,glob){
    };
    ~CircleFilter(){
    
    };
    
    void get_extended_params(){
    };
    
    IplImage* image_filter(IplImage* img_ptr,sensor_msgs::CameraInfo* cam_info_ptr){
        return (IplImage*) img_ptr;
    };
};


int main(int argc, char** argv)
{
    ros::init(argc, argv, "snapshot_filter");
    ROS_INFO("Main Called");
    ros::NodeHandle n("~");
    ros::NodeHandle glob;
    ROS_INFO("NH instantiated");
    SnapshotFilter ic(n,glob);
    ROS_INFO("Successfully created snapshotfilter");
    ros::spin();
    return 0;
}
#endif

